numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -385,7 +358,10 @@ string_rfind_loop(PyArrayMethod_Context *context,
     while (N--) {
         Buffer<enc> buf1(in1, elsize1);
         Buffer<enc> buf2(in2, elsize2);
-        npy_intp idx = string_rfind<enc>(buf1, buf2, *(npy_int64 *)in3, *(npy_int64 *)in4);
+        npy_intp idx = function(buf1, buf2, *(npy_int64 *)in3, *(npy_int64 *)in4);
+        if (PyErr_Occurred()) {
```

Comment:
I don't quite know whether `PyErr_Occurred()` is fast or slow, but in principle one could add an additional bit of data to `static_data` that tells whether this test is necessary. Alternatively, maybe that is still a good reason for expanding the template to `<ENCODING end, bool can_py_err>`?

On the other hand, even counting is probably slow compared to checking for the error; surely this is optimized quite well by python! So, definitely fine also to leave as is.

Reply:
Came from an email:  Do you even hold the GIL to check (which is needed)?  Although I thought python debug would flush such an error out.

Reply:
> Do you even hold the GIL to check (which is needed)?

Oh, that's a good point. Maybe it indeed makes sense to just have an additional bit of info in `static_data`.

Reply:
Sorry, but that doesn't work: the reason why `static_data` is called `static` :).  

That is why I thought adding some `scratch_data` to the `context` is likely useful (but I didn't have a pressing enough need to actually add it yet, and I always tried to be minimal).  Which you can feel free to do:  A uint64 or two with the only guarantee that NumPy will NULL it before the first call.

Reply:
In case it wasn't clear: The reason is that using that space wouldn't be thread-safe.

Reply:
It does work - the static data can tell whether you are in `index` (which can do `npy_gil` stuff) or in regular `find` (which cannot).

But I do think this points to the larger problem - can one check `npy_gil...` has been called in the function?

Reply:
Ah, sorry, I thought you needed to keep track of whether an error occurred or not.  The nicest thing would be transition to new loops and just fail directly.

Reply:
I ended up going with just returning `-2` from `index`/`rindex` as a sentinel for exception raised. Does that work?